airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py

```diff
@@ -0,0 +1,467 @@
+        :return: reference to relevant class
+        :rtype: type
+        """
+
+    @staticmethod
+    @abstractmethod
+    def filepath_iterator(logger: AirbyteLogger, provider: dict) -> Iterator[str]:
+        """
+        Provider-specific method to iterate through bucket/container/etc. and yield each full filepath.
+        This should supply the 'url' to use in FileClient(). This is possibly better described as blob or file path.
+            e.g. for AWS: f"s3://{aws_access_key_id}:{aws_secret_access_key}@{self._url}" <- self._url is what we want to yield here
+
+        :param logger: instance of AirbyteLogger to use as this is a staticmethod
+        :type logger: AirbyteLogger
+        :param provider: provider specific mapping as described in spec.json
+        :type provider: dict
+        :yield: url filepath to use in FileClient()
+        :rtype: Iterator[str]
+        """
+
+    def pattern_matched_filepath_iterator(self, filepaths:Iterable[str]) -> Iterator[str]:
+        """
+        iterates through iterable filepaths and yields only those filepaths that match user-provided path patterns
+
+        :param filepaths: filepath_iterator(), this is a param rather than method reference in order to unit test this
+        :type filepaths: Iterable[str]
+        :yield: url filepath to use in FileClient(), if matching on user-provided path patterns
+        :rtype: Iterator[str]
+        """
+        for filepath in filepaths:
+            for path_pattern in self._path_patterns:
+                if fnmatch(filepath, path_pattern):
+                    yield filepath
+
+    def time_ordered_fileclient_iterator(self) -> Iterable[Tuple[datetime, FileClient]]:
+        """
+        Iterates through pattern_matched_filepath_iterator(), acquiring last_modified property of each file to return in time ascending order.
+        Uses concurrent.futures to thread this asynchronously in order to improve performance when there are many files (network I/O)
+        Caches results after first run of method to avoid repeating network calls as this is used more than once 
+
+        :return: list of tuples in format ({last modified date}, {FileClient object})
+        :rtype: Iterable[Tuple[datetime, FileClient]]
+        """
+
+        def get_fileclient_with_lastmod(filepath: str) -> Tuple[datetime, FileClient]:
+            fc = self.fileclient_class(filepath, self._provider)
+            return (fc.last_modified, fc)
+
+        if self.fileclient_cache is None:
+            fileclients = []  # list of tuples (datetime, FileClient) which we'll use to sort
+            # use concurrent future threads to parallelise grabbing last_modified from all the files
+            # TODO: don't hardcode max_workers like this
+            with concurrent.futures.ThreadPoolExecutor(max_workers=64) as executor:
+                futures = {
+                    executor.submit(get_fileclient_with_lastmod, fp): 
```

Comment:
why are we passing a dict of `{future: path}` into `as_completed` when a list would do? I might have missed this but I was trying to figure out why this is a dict when `as_completed` takes an input sequence of futures, and the fact that this works happens only because iterating over a `dict` iterates over its keys by default 

![Screen Shot 2021-07-26 at 10 08 58 PM](https://user-images.githubusercontent.com/6246757/127098831-82944684-af4e-4d12-84af-1a90afcafece.png)

Reply:
good point, dict is redundant here, changed to list